### PR TITLE
HDDS-8668. Allow reconfiguration of annotated config object

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

 * allow `@Config` properties to be marked as `reconfigurable`
 * implement reconfiguration of annotated config object from `ConfigurationSource`
 * create base class `ReconfigurableConfig` to help with keeping track of reconfigurable properties and handling updates

This is not tied into reconfiguration handling in OM/SCM/etc., will be done in follow-up.

https://issues.apache.org/jira/browse/HDDS-8668

## How was this patch tested?

Added unit tests.

CI:
https://github.com/adoroszlai/hadoop-ozone/actions/runs/5073256233

### CODE REVIEW

**Comment by adoroszlai** [comment]
@whbing @xichen01 please review when you have some time

**Review by xichen01 — COMMENTED** [review]
It looks good overall, this is a useful feature. I have left few comments you can refer to.

File: hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationReflectionUtil.java [hunk 1]
```diff
@@ -34,99 +36,101 @@ public final class ConfigurationReflectionUtil {
...
         Config configAnnotation = field.getAnnotation(Config.class);
 
-        String key = prefix + "." + configAnnotation.key();
+        if (reconfiguration && !configAnnotation.reconfigurable()) {
+          continue;
+        }
 
+        String key = prefix + "." + configAnnotation.key();
         String defaultValue = configAnnotation.defaultValue();
+        String value = from.get(key, defaultValue);
```
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationReflectionUtil.java:100]: The config object may have other value (e.g. if conf had non-default value previously), so we need to set it anyway.  We may skip setting it if the field's actual value is the same as the new value, but I don't think it's worth adding the equality check.

File: hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationReflectionUtil.java [hunk 2]
```diff
@@ -34,99 +36,101 @@ public final class ConfigurationReflectionUtil {
...
 
-        ConfigType type = configAnnotation.type();
+        setField(configurationClass, configuration, field, configAnnotation,
```
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationReflectionUtil.java:102]: Correct, I think none of the type-safe config objects in Ozone are thread-safe, but it's not specific to this change.

**Comment by whbing** [comment]
The code is elegant and generic !  LGTM +1

File: hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationReflectionUtil.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (xichen01) [reply hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationReflectionUtil.java:100]: OK, understood

File: hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationReflectionUtil.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (xichen01) [reply hadoop-hdds/config/src/main/java/org/apache/hadoop/hdds/conf/ConfigurationReflectionUtil.java:102]: So it is possible that some synchronization operations may be required when implementing specific configuration updates.

**Comment by xichen01** [comment]
@adoroszlai I have a curious question about the future code implementation.

What is the relationship between configurations with the `@Config(reconfigurable = true)` annotation and configurations supported by the `reconfig` command?
1. Are all configurations with the `@Config(reconfigurable = true)` annotation supported by the `reconfig` command
2. Need to manually implemented which configurations are supported by the `reconfig` command.

**Comment by adoroszlai** [comment]
> What is the relationship between configurations with the `@Config(reconfigurable = true)` annotation and configurations supported by the `reconfig` command?

Currently there is no relationship yet.  We'll need to register the config objects with the one that handles `reconfig` commands.

**Review by symious — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @symious, @whbing, @xichen01 for the review.

**Pull request merged**